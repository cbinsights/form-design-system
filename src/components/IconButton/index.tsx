import React, { forwardRef } from 'react';
import cc from 'classcat';
import baseElement from 'util/baseElement';
import { IconProps } from 'components/Icon';

export const THEMES = ['ghost', 'aqua', 'outlined'];
export const RADII = ['square', 'circle'];
export const SIZES = ['s', 'm'];

export interface IconButtonProps {
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link?: React.ElementType;
  /** Controls active style UI of button */
  isActive?: boolean;
  /** Controls radius of button (slightly rounded square, or circle) */
  radius?: 'square' | 'circle';
  /** Controls spinner showing for icon button (normal icon is hidden) */
  isLoading?: boolean;
  /** Controls look and feel of button */
  theme?: 'ghost' | 'aqua' | 'outlined';
  /** Controls destructive look and feel of button */
  isDestructive?: boolean;
  /** Used to control the size of the button */
  size?: 's' | 'm';
  /** Controls whether the button is disabled or not. */
  disabled?: boolean;
  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon: React.ComponentType<IconProps>;
  /** Accessibility label */
  label: string;
  /** Extend click radius of button to nearest relative parent */
  isBreakoutLink?: boolean;
  /** URL for base element */
  href?: string;
  /** Handler for on click event */
  onClick?: () => void;
}

const IconButton = forwardRef(
  (
    {
      radius = 'square',
      theme = 'ghost',
      size = 'm',
      Icon,
      disabled,
      isActive,
      isLoading,
      isDestructive,
      label,
      Link,
      isBreakoutLink,
      ...rest
    }: IconButtonProps,
    ref
  ) => {
    const Element = baseElement({ href: rest.href, onClick: true, as: Link });
    return (
      <Element
        {...rest}
        ref={ref}
        title={label}
        className={cc([
          {
            'shape--circle': radius === 'circle',
            'fdsIconButton--disabled': disabled,
            'fdsIconButton--isActive': isActive && !disabled,
            'fdsIconButton--isDestructive': isDestructive,
            'fdsIconButton--loading': isLoading,
            breakoutLink: isBreakoutLink,
          },
          'fdsIconButton',
          'rounded--all',
          'border--focus--noTransition',
          'transition--default',
          'alignChild--center--center',
          `fdsIconButton--${theme}`,
          `fdsIconButton--${size}`,
        ])}
        disabled={disabled && Element === 'button'}
      >
        <div className="fdsIconButton--loading-spinnerWrapper">
          <div className={cc({ 'fdsIconButton--loading-spinner': isLoading })} />
        </div>
        <span className={isLoading ? 'fdsIconButton--hidden' : ''}>
          <Icon customSize={size === 's' ? 16 : 18} />
        </span>
      </Element>
    );
  }
);

IconButton.displayName = 'IconButton';

export default IconButton;
