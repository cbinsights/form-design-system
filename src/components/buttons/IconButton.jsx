import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from '../../util/baseElement';

export const THEMES = ['aqua', 'ghost'];
export const RADII = ['square', 'circle'];
export const SIZES = ['s', 'm'];

const IconButton = ({
  Icon,
  disabled,
  isActive,
  radius,
  theme,
  isLoading,
  size,
  isDestructive,
  label,
  Link,
  ...rest
}) => {
  const Element = baseElement({ href: rest.href, onClick: true, as: Link });
  return (
    <Element
      {...rest}
      title={label}
      className={cx(
        'fdsIconButton',
        'rounded--all',
        'border--focus--noTransition',
        'transition-default',
        `fdsIconButton--${theme}`,
        `fdsIconButton--${size}`,
        {
          'fdsIconButton--disabled': disabled,
          'fdsIconButton--active': isActive,
          'fdsIconButton--circle': radius === 'circle',
          'fdsIconButton--isDestructive': isDestructive,
          'fdsIconButton--loading': isLoading,
        }
      )}
      disabled={disabled && Element === 'button'}
    >
      <span className={isLoading ? 'fdsIconButton--hidden' : ''}>
        <Icon customSize={size === 's' ? 16 : 18} />
      </span>
    </Element>
  );
};

IconButton.defaultProps = {
  radius: 'square',
  theme: 'ghost',
  size: 'm',
};

IconButton.propTypes = {
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
  /** Controls active style UI of button */
  isActive: PropTypes.bool,
  /** Controls radius of button (slightly rounded square, or circle) */
  radius: PropTypes.oneOf(RADII),
  /** Controls spinner showing for icon button (normal icon is hidden) */
  isLoading: PropTypes.bool,
  /** Controls look and feel of button */
  theme: PropTypes.oneOf(THEMES),
  /** Controls destructive look and feel of button */
  isDestructive: PropTypes.bool,
  /** Used to control the size of the button */
  size: PropTypes.oneOf(SIZES),
  /** Controls whether the button is disabled or not. */
  disabled: PropTypes.bool,
  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon: PropTypes.func,
  /** Accessibility label */
  label: PropTypes.string,
};

export default IconButton;
