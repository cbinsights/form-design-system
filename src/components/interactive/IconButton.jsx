import React from 'react';
import PropTypes from 'prop-types';
import cc from 'classcat';
import baseElement from 'util/baseElement';

export const THEMES = ['ghost', 'aqua', 'outlined'];
export const RADII = ['square', 'circle'];
export const SIZES = ['s', 'm'];

const IconButton = ({
  radius = 'square',
  theme = 'ghost',
  size = 'm',
  Icon,
  disabled,
  isActive,
  isLoading,
  isDestructive,
  label,
  Link,
  ...rest
}) => {
  const Element = baseElement({ href: rest.href, onClick: true, as: Link });
  return (
    <Element
      {...rest}
      title={label}
      className={cc([
        {
          'shape--circle': radius === 'circle',
          'fdsIconButton--disabled': disabled,
          'fdsIconButton--isActive': isActive && !disabled,
          'fdsIconButton--isDestructive': isDestructive,
          'fdsIconButton--loading': isLoading,
        },
        'fdsIconButton',
        'rounded--all',
        'border--focus--noTransition',
        'transition--default',
        'alignChild--center--center',
        `fdsIconButton--${theme}`,
        `fdsIconButton--${size}`,
      ])}
      disabled={disabled && Element === 'button'}
    >
      <span className={isLoading ? 'fdsIconButton--hidden' : ''}>
        <Icon customSize={size === 's' ? 16 : 18} />
      </span>
    </Element>
  );
};

IconButton.propTypes = {
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
  /** Controls active style UI of button */
  isActive: PropTypes.bool,
  /** Controls radius of button (slightly rounded square, or circle) */
  radius: PropTypes.oneOf(RADII),
  /** Controls spinner showing for icon button (normal icon is hidden) */
  isLoading: PropTypes.bool,
  /** Controls look and feel of button */
  theme: PropTypes.oneOf(THEMES),
  /** Controls destructive look and feel of button */
  isDestructive: PropTypes.bool,
  /** Used to control the size of the button */
  size: PropTypes.oneOf(SIZES),
  /** Controls whether the button is disabled or not. */
  disabled: PropTypes.bool,
  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon: PropTypes.func.isRequired,
  /** Accessibility label */
  label: PropTypes.string.isRequired,
};

export default IconButton;
