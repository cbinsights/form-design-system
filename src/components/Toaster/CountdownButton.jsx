import React from 'react';
import PropTypes from 'prop-types';
import { Transition } from 'react-transition-group';
import FDS from 'lib/dictionary/js/styleConstants';
import IconButton from 'components/IconButton';
import DenyIcon from 'lib/icons/react/DenyIcon';

/**
 *
 * @param {Number} circleSize bounding box size of circle in px
 * @param {Number} strokeWidth width of circle stroke
 */
export const getCircleInfo = (circleSize, strokeWidth) => {
  const radius = (circleSize - strokeWidth) / 2;

  return {
    r: radius,
    c: Math.PI * radius * 2,
    centerOffset: circleSize / 2,
  };
};

const CountdownButton = ({ label = 'Close', Icon = DenyIcon, duration, onClick }) => {
  const circleSize = 34;
  const strokeWidth = '2';
  const { r, c, centerOffset } = getCircleInfo(circleSize, strokeWidth);

  const defaultCircleStyle = {
    // strokeDasharray sets up a dashed line in which each dash is equal to
    // the circumference of the circle.
    //
    // We offset the dash by 0 to start to create a full circle
    strokeDasharray: c,
    strokeDashoffset: 0,
    transition: `all ${duration}ms linear`,

    // start the countdown at "midnight" position of circle
    transformOrigin: 'center center',
    transform: 'rotate(-90deg)',
  };

  const transitionStyles = {
    entering: { strokeDashoffset: 0 },

    // offsetting the dash by the exact circle circumference lines up
    // the empty part of the dashed stroke with the circle circumference
    entered: { strokeDashoffset: c },
  };

  return (
    <div
      className="countdownButton"
      style={{
        width: `${circleSize}px`,
        height: `${circleSize}px`,
      }}
    >
      <div className="countdownButton-button">
        <IconButton onClick={onClick} radius="circle" Icon={Icon} label={label} />
      </div>
      <Transition
        in={true}
        appear
        timeout={{
          appear: 1,
          enter: duration,
        }}
        mountOnEnter
      >
        {(state) => (
          <svg
            className="countdownButton-animation"
            width={circleSize}
            height={circleSize}
            viewBox={`0 0 ${circleSize} ${circleSize}`}
          >
            <circle
              cx={centerOffset}
              cy={centerOffset}
              r={r}
              strokeWidth={strokeWidth}
              fill="transparent"
              stroke={FDS.FONT_COLOR_HINT}
              style={{
                ...defaultCircleStyle,
                ...transitionStyles[state],
              }}
            />
          </svg>
        )}
      </Transition>
    </div>
  );
};

CountdownButton.propTypes = {
  /** Duration in ms of countdown animation */
  duration: PropTypes.number.isRequired,

  /** Used to render a FDS Icon (should only be used for FDS Icons, e.g. Icon={DenyIcon}) */
  Icon: PropTypes.func,

  /** adds a label to IconButton */
  label: PropTypes.string,

  /** Click callback for button */
  onClick: PropTypes.func,
};

export default CountdownButton;
