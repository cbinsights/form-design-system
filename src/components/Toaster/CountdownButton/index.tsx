import React from 'react';
import { Transition } from 'react-transition-group';
import FDS from 'dictionary/js/styleConstants';
import IconButton from 'components/IconButton';
import DenyIcon from 'icons/react/DenyIcon';
import Icon from 'components/Icon';
export interface CircleInfo {
  radius: number;
  circumference: number;
  centerOffset: number;
}

/**
 *
 * @param {Number} circleSize bounding box size of circle in px
 * @param {Number} strokeWidth width of circle stroke
 */
export const getCircleInfo = (circleSize: number, strokeWidth: number): CircleInfo => {
  const radius = (circleSize - strokeWidth) / 2;

  return {
    radius,
    circumference: Math.PI * radius * 2,
    centerOffset: circleSize / 2,
  };
};

export interface CountdownButtonProps {
  /** Duration in ms of countdown animation */
  duration: number;

  /** Used to render a FDS Icon (should only be used for FDS Icons, e.g. Icon={DenyIcon}) */
  Icon?: Icon;

  /** adds a label to IconButton */
  label?: string;

  /** Click callback for button */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
}

const CountdownButton = ({
  label = 'Close',
  Icon = DenyIcon,
  duration,
  onClick,
}: CountdownButtonProps): JSX.Element => {
  const circleSize = 34;
  const strokeWidth = 2;
  const { radius, circumference, centerOffset } = getCircleInfo(circleSize, strokeWidth);

  const defaultCircleStyle = {
    // strokeDasharray sets up a dashed line in which each dash is equal to
    // the circumference of the circle.
    //
    // We offset the dash by 0 to start to create a full circle
    strokeDasharray: circumference,
    strokeDashoffset: 0,
    transition: `all ${duration}ms linear`,

    // start the countdown at "midnight" position of circle
    transformOrigin: 'center center',
    transform: 'rotate(-90deg)',
  };

  const transitionStyles = {
    entering: { strokeDashoffset: 0 },

    // offsetting the dash by the exact circle circumference lines up
    // the empty part of the dashed stroke with the circle circumference
    entered: { strokeDashoffset: circumference },
    exiting: undefined,
    exited: undefined,
    unmounted: undefined,
  };

  return (
    <div
      className="countdownButton"
      style={{
        width: `${circleSize}px`,
        height: `${circleSize}px`,
      }}
    >
      <div className="countdownButton-button">
        <IconButton onClick={onClick} radius="circle" Icon={Icon} label={label} />
      </div>
      <Transition
        in={true}
        appear
        timeout={{
          appear: 1,
          enter: duration,
        }}
        mountOnEnter
      >
        {(state) => (
          <svg
            className="countdownButton-animation"
            width={circleSize}
            height={circleSize}
            viewBox={`0 0 ${circleSize} ${circleSize}`}
          >
            <circle
              cx={centerOffset}
              cy={centerOffset}
              r={radius}
              strokeWidth={strokeWidth}
              fill="transparent"
              stroke={FDS.FONT_COLOR_HINT}
              style={{
                ...defaultCircleStyle,
                ...transitionStyles[state],
              }}
            />
          </svg>
        )}
      </Transition>
    </div>
  );
};

export default CountdownButton;
