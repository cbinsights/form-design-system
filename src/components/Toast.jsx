import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import IconButton from './buttons/IconButton';
import Button from './buttons/Button';
import CountdownButton from './buttons/CountdownButton';
import Flex from './layout/Flex';
import FlexItem from './layout/FlexItem';
import customPropTypes from './util/customPropTypes';
import FDS from '../../lib/dictionary/js/styleConstants';
import CheckIcon from '../../lib/icons/react/CheckIcon';
import InformationIcon from '../../lib/icons/react/InformationIcon';
import DenyIcon from '../../lib/icons/react/DenyIcon';

const Toast = ({
  content,
  type,
  actionLabel,
  progress,
  onAction,
  isAutoDismiss,
  dismissDelay,
  canDismiss,
  dismissToast /* Passed from Toaster */,
}) => {
  const classNames = cx('toast', `toast--${type}`);

  const icon = {
    info: <InformationIcon size="l" color={FDS.COLOR_AQUA} />,
    success: <CheckIcon size="xs" color={FDS.COLOR_WHITE} />,
    warn: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    error: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
  }[type];

  const useContinuousProgressBar = type === 'progress' && !progress;
  const useAmountProgressBar = type === 'progress' && progress;

  const toastBar = (
    <div className="toast-bar">
      {useContinuousProgressBar && (
        <div className="toast-progressBar toast-progressBar--continuous" />
      )}
      {useAmountProgressBar && (
        <div
          className="toast-progressBar toast-progressBar--amount"
          style={{ width: `${progress}%` }}
        />
      )}
    </div>
  );

  const onActionDismiss = (event) => {
    dismissToast();
    if (typeof onAction === 'function') {
      onAction(event);
    }
  };

  return (
    <div
      className={classNames}
      role={type === 'error' ? 'alert' : 'log'}
      aria-relevant="all"
      aria-atomic="true"
    >
      {toastBar}
      <div className="toast-content alignChild--left--center">
        <Flex>
          {icon && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <div className="toast-icon media--s">{icon}</div>
              </div>
            </FlexItem>
          )}
          <FlexItem>
            <div className="toast-constrainGrowth">{content}</div>
          </FlexItem>
          {actionLabel && onAction && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <Button onClick={onActionDismiss} data-test="toast-action" theme="ghost">
                  {actionLabel}
                </Button>
              </div>
            </FlexItem>
          )}
          {canDismiss && type !== 'progress' && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                {isAutoDismiss ? (
                  <CountdownButton onClick={dismissToast} duration={dismissDelay} />
                ) : (
                  <IconButton Icon={DenyIcon} onClick={dismissToast} radius="circle" />
                )}
              </div>
            </FlexItem>
          )}
        </Flex>
      </div>
    </div>
  );
};

Toast.defaultProps = {
  dismissDelay: 4000,
  canDismiss: true,
  isAutoDismiss: true,
  onDismiss: () => {},
};

Toast.propTypes = {
  /** JSX Content of Toast */
  content: PropTypes.element.isRequired,

  /** Type of toast */
  type: PropTypes.oneOf(['info', 'warn', 'error', 'success', 'progress']).isRequired,

  /** Label for action button */
  actionLabel: PropTypes.string,

  /** Callback for action button click */
  onAction: PropTypes.func,

  /** Callback invoked on dismissal of toast */
  onDismiss: PropTypes.func,

  /** Time in ms to auto-dismiss toast */
  dismissDelay: PropTypes.number,

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,

  /** Is this toast user-dismissable? */
  canDismiss: PropTypes.bool,

  /** Number 1-100 declaring % progress */
  progress: customPropTypes.range(1, 100),

  /** Passed by Toaster: function to dismiss the toast */
  dismissToast: PropTypes.func,
};

export default Toast;
