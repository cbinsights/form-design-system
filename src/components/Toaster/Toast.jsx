import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import FDS from 'lib/dictionary/js/styleConstants';
import CheckIcon from 'lib/icons/react/CheckIcon';
import InformationIcon from 'lib/icons/react/InformationIcon';
import DenyIcon from 'lib/icons/react/DenyIcon';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import customPropTypes from 'components/util/customPropTypes';
import Button from 'components/Button';
import IconButton from 'components/IconButton';

import CountdownButton from './CountdownButton';

export const TYPES = ['info', 'warn', 'error', 'success', 'progress'];

const Toast = ({
  isAutoDismiss = true,
  dismissDelay = 4000,
  canDismiss = true,
  dismissOnAction = true,
  type = 'info',
  content,
  actionLabel,
  progress,
  onAction,
  dismissToast /* Passed from Toaster */,
}) => {
  const classNames = cx('toast', `toast--${type}`);

  const icon = {
    info: <InformationIcon size="l" color={FDS.COLOR_AQUA} />,
    success: <CheckIcon size="xs" color={FDS.COLOR_WHITE} />,
    warn: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    error: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
  }[type];

  const useContinuousProgressBar = type === 'progress' && !progress;
  const useAmountProgressBar = type === 'progress' && progress;

  const toastBar = (
    <div className="toast-bar">
      {useContinuousProgressBar && (
        <div className="toast-progressBar toast-progressBar--continuous" />
      )}
      {useAmountProgressBar && (
        <div
          className="toast-progressBar toast-progressBar--amount"
          style={{ width: `${progress}%` }}
        />
      )}
    </div>
  );

  const onActionDismiss = (event) => {
    if (dismissOnAction) dismissToast();
    if (onAction) onAction(event);
  };

  return (
    <div
      className={classNames}
      role={type === 'error' ? 'alert' : 'log'}
      aria-relevant="all"
      aria-atomic="true"
    >
      {toastBar}
      <div className="toast-content alignChild--left--center">
        <Flex>
          {icon && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <div className="toast-icon media--s">{icon}</div>
              </div>
            </FlexItem>
          )}
          <FlexItem>
            <div className="toast-constrainGrowth">{content}</div>
          </FlexItem>
          {actionLabel && (onAction || (dismissToast && dismissOnAction)) && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <Button
                  onClick={onActionDismiss}
                  data-test="toast-action"
                  theme="ghost"
                  label={actionLabel}
                />
              </div>
            </FlexItem>
          )}
          {canDismiss && type !== 'progress' && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                {isAutoDismiss ? (
                  <CountdownButton
                    onClick={dismissToast}
                    duration={dismissDelay}
                    label="Close"
                  />
                ) : (
                  <IconButton
                    Icon={DenyIcon}
                    onClick={dismissToast}
                    radius="circle"
                    label="Close"
                  />
                )}
              </div>
            </FlexItem>
          )}
        </Flex>
      </div>
    </div>
  );
};

Toast.propTypes = {
  /** JSX Content of Toast */
  content: PropTypes.node.isRequired,

  /** Type of toast */
  type: PropTypes.oneOf(TYPES),

  /** Label for action button */
  actionLabel: PropTypes.string,

  /** Callback for action button click */
  onAction: PropTypes.func,

  /** Time in ms to auto-dismiss toast */
  dismissDelay: PropTypes.number,

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,

  /* Controls if toast gets dismiss when the ActionButton is clicked */
  dismissOnAction: PropTypes.bool,

  /** Is this toast user-dismissable? */
  canDismiss: PropTypes.bool,

  /** Number 1-100 declaring % progress */
  progress: customPropTypes.range(1, 100),

  /** Passed by Toaster: function to dismiss the toast */
  dismissToast: PropTypes.func,
};

export default Toast;
