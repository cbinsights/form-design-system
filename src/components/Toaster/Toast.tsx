import React from 'react';
import cc from 'classcat';
import FDS from 'dictionary/js/styleConstants';
import CheckIcon from 'icons/react/CheckIcon';
import InformationIcon from 'icons/react/InformationIcon';
import DenyIcon from 'icons/react/DenyIcon';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import Button from 'components/Button';
import IconButton from 'components/IconButton';

import CountdownButton from './CountdownButton';

type ToastType = 'info' | 'warn' | 'error' | 'success' | 'progress';

export interface ToastProps {
  /** JSX Content of Toast */
  content: React.ReactNode;

  /** Type of toast */
  type?: ToastType;

  /** Label for action button */
  actionLabel?: string;

  /** Callback for action button click */
  onAction?: React.MouseEventHandler<HTMLButtonElement>;

  /** Time in ms to auto-dismiss toast */
  dismissDelay?: number;

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss?: boolean;

  /* Controls if toast gets dismiss when the ActionButton is clicked */
  dismissOnAction?: boolean;

  /** Is this toast user-dismissable? */
  canDismiss?: boolean;

  /** Number 1-100 declaring % progress */
  progress?: number;

  /** Passed by Toaster: function to dismiss the toast */
  dismissToast?(...args: unknown[]): unknown;
}

const Toast = ({
  isAutoDismiss = true,
  dismissDelay = 4000,
  canDismiss = true,
  dismissOnAction = true,
  type = 'info',
  content,
  actionLabel,
  progress,
  onAction,
  dismissToast /* Passed from Toaster */,
}: ToastProps): JSX.Element => {
  const classNames = cc(['toast', `toast--${type}`]);

  const icon = {
    info: <InformationIcon size="l" color={FDS.COLOR_WHITE} />,
    success: <CheckIcon size="xs" color={FDS.COLOR_WHITE} />,
    warn: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    error: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    progress: undefined,
  }[type];

  const accentColor = {
    info: FDS.COLOR_BLUE,
    warn: FDS.COLOR_ORANGE,
    error: FDS.COLOR_RED,
    success: FDS.COLOR_GREEN,
    progress: FDS.COLOR_AQUA,
  }[type];

  const iconStyle = { backgroundColor: accentColor };

  const onActionDismiss: React.MouseEventHandler<HTMLButtonElement> = (event) => {
    if (dismissOnAction && dismissToast) dismissToast();
    if (onAction) onAction(event);
  };

  return (
    <div
      className={classNames}
      role={type === 'error' ? 'alert' : 'log'}
      aria-relevant="all"
      aria-atomic="true"
    >
      <div className="toast-bar">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          className={cc([
            'toast-svg',
            { 'toast-svg--animate': type === 'progress' && !progress },
          ])}
        >
          <line
            x1="0%"
            y1="50%"
            x2={type === 'progress' && progress ? `${progress}%` : '100%'}
            y2="50%"
            stroke={accentColor}
            strokeWidth={FDS.SPACE_S}
            strokeDasharray={type === 'progress' && !progress ? '200 400' : ''}
          />
        </svg>
      </div>
      <div className="toast-content alignChild--left--center">
        <Flex>
          {icon && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <div className="toast-icon media--s" style={iconStyle}>
                  {icon}
                </div>
              </div>
            </FlexItem>
          )}
          <FlexItem>
            <div className="toast-constrainGrowth">{content}</div>
          </FlexItem>
          {actionLabel && (onAction || (dismissToast && dismissOnAction)) && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <Button
                  onClick={onActionDismiss}
                  data-test="toast-action"
                  theme="ghost"
                  label={actionLabel}
                />
              </div>
            </FlexItem>
          )}
          {canDismiss && type !== 'progress' && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                {isAutoDismiss ? (
                  <CountdownButton
                    onClick={dismissToast}
                    duration={dismissDelay}
                    label="Close"
                  />
                ) : (
                  <IconButton
                    Icon={DenyIcon}
                    onClick={dismissToast}
                    radius="circle"
                    label="Close"
                  />
                )}
              </div>
            </FlexItem>
          )}
        </Flex>
      </div>
    </div>
  );
};

export default Toast;
