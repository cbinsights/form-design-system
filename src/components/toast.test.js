import React from 'react';
import { shallow } from 'enzyme';

import IconButton from './IconButton';
import CountdownButton from '../media/CountdownButton';

import Toast from './Toast';

const renderByType = (type, extraProps) =>
  shallow(<Toast type={type} content={<p>lol</p>} {...extraProps} />);


describe('Toast component', () => {

  it('matches snapshot (info)', () => {
    const component = renderByType('info');
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (success)', () => {
    const component = renderByType('success');
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (progress)', () => {
    const component = renderByType('progress');
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (error)', () => {
    const component = renderByType('error');
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (warn)', () => {
    const component = renderByType('warn');
    expect(component).toMatchSnapshot();
  });

  it('does NOT render action button if label and action are not passed', () => {
    const component = renderByType('info');
    const actionButton = component.find('[data-test="toast-action"]');
    expect(actionButton).toHaveLength(0);
  });

  it('renders action button if label and action are passed', () => {
    const component = renderByType('info', {
      onAction: () => {},
      actionLabel: 'action',
    });
    const actionButton = component.find('[data-test="toast-action"]');
    expect(actionButton).toHaveLength(1);
  });

  it('does NOT render close button when `canDismiss` set to false', () => {
    const component = renderByType('info', {
      canDismiss: false,
    });
    const closeButton = component.find(CountdownButton);
    expect(closeButton).toHaveLength(0);
  });

  it('does NOT render close button for "progress" toasts', () => {
    const component = renderByType('progress');
    const closeButton = component.find(CountdownButton);
    expect(closeButton).toHaveLength(0);
  });

 it('renders close button by default', () => {
    const component = renderByType('info');
    const closeButton = component.find(CountdownButton);
    expect(closeButton).toHaveLength(1);
  });

  it('uses normal dismiss button when `isAutoDismiss` set to false', () => {
    const component = renderByType('info', {
      isAutoDismiss: false,
    });

    const countdownButton = component.find(CountdownButton);
    const standardButton = component.find(IconButton);
    expect(countdownButton).toHaveLength(0);
    expect(standardButton).toHaveLength(1);
  });
});
