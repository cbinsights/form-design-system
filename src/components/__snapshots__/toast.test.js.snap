// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast component matches snapshot (error) 1`] = `
<div
  aria-atomic="true"
  aria-relevant="all"
  className="toast toast--error"
  role="alert"
>
  <div
    className="toast-bar"
  />
  <div
    className="toast-content alignChild--left--center"
  >
    <Flex
      align="stretch"
      direction="row"
    >
      <FlexItem
        shrink={true}
      >
        <div
          className="toast-constrainGrowth alignChild--center--center"
        >
          <div
            className="toast-icon media--s"
          >
            <div
              className="fontWeight--bold color--white"
              role="image"
            >
              !
            </div>
          </div>
        </div>
      </FlexItem>
      <FlexItem
        shrink={false}
      >
        <div
          className="toast-constrainGrowth"
        >
          <p>
            lol
          </p>
        </div>
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <div
          className="toast-constrainGrowth alignChild--center--center"
        >
          <CountdownButton
            Icon={[Function]}
            duration={4000}
          />
        </div>
      </FlexItem>
    </Flex>
  </div>
</div>
`;

exports[`Toast component matches snapshot (info) 1`] = `
<div
  aria-atomic="true"
  aria-relevant="all"
  className="toast toast--info"
  role="log"
>
  <div
    className="toast-bar"
  />
  <div
    className="toast-content alignChild--left--center"
  >
    <Flex
      align="stretch"
      direction="row"
    >
      <FlexItem
        shrink={true}
      >
        <div
          className="toast-constrainGrowth alignChild--center--center"
        >
          <div
            className="toast-icon media--s"
          >
            <InformationIcon
              color="#4cabce"
              size="l"
            />
          </div>
        </div>
      </FlexItem>
      <FlexItem
        shrink={false}
      >
        <div
          className="toast-constrainGrowth"
        >
          <p>
            lol
          </p>
        </div>
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <div
          className="toast-constrainGrowth alignChild--center--center"
        >
          <CountdownButton
            Icon={[Function]}
            duration={4000}
          />
        </div>
      </FlexItem>
    </Flex>
  </div>
</div>
`;

exports[`Toast component matches snapshot (progress) 1`] = `
<div
  aria-atomic="true"
  aria-relevant="all"
  className="toast toast--progress"
  role="log"
>
  <div
    className="toast-bar"
  >
    <div
      className="toast-progressBar toast-progressBar--continuous"
    />
  </div>
  <div
    className="toast-content alignChild--left--center"
  >
    <Flex
      align="stretch"
      direction="row"
    >
      <FlexItem
        shrink={false}
      >
        <div
          className="toast-constrainGrowth"
        >
          <p>
            lol
          </p>
        </div>
      </FlexItem>
    </Flex>
  </div>
</div>
`;

exports[`Toast component matches snapshot (success) 1`] = `
<div
  aria-atomic="true"
  aria-relevant="all"
  className="toast toast--success"
  role="log"
>
  <div
    className="toast-bar"
  />
  <div
    className="toast-content alignChild--left--center"
  >
    <Flex
      align="stretch"
      direction="row"
    >
      <FlexItem
        shrink={true}
      >
        <div
          className="toast-constrainGrowth alignChild--center--center"
        >
          <div
            className="toast-icon media--s"
          >
            <CheckIcon
              color="#ffffff"
              size="xs"
            />
          </div>
        </div>
      </FlexItem>
      <FlexItem
        shrink={false}
      >
        <div
          className="toast-constrainGrowth"
        >
          <p>
            lol
          </p>
        </div>
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <div
          className="toast-constrainGrowth alignChild--center--center"
        >
          <CountdownButton
            Icon={[Function]}
            duration={4000}
          />
        </div>
      </FlexItem>
    </Flex>
  </div>
</div>
`;

exports[`Toast component matches snapshot (warn) 1`] = `
<div
  aria-atomic="true"
  aria-relevant="all"
  className="toast toast--warn"
  role="log"
>
  <div
    className="toast-bar"
  />
  <div
    className="toast-content alignChild--left--center"
  >
    <Flex
      align="stretch"
      direction="row"
    >
      <FlexItem
        shrink={true}
      >
        <div
          className="toast-constrainGrowth alignChild--center--center"
        >
          <div
            className="toast-icon media--s"
          >
            <div
              className="fontWeight--bold color--white"
              role="image"
            >
              !
            </div>
          </div>
        </div>
      </FlexItem>
      <FlexItem
        shrink={false}
      >
        <div
          className="toast-constrainGrowth"
        >
          <p>
            lol
          </p>
        </div>
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <div
          className="toast-constrainGrowth alignChild--center--center"
        >
          <CountdownButton
            Icon={[Function]}
            duration={4000}
          />
        </div>
      </FlexItem>
    </Flex>
  </div>
</div>
`;
